### ISSUE TITLE
Punctuation should be ignored in all places and ways in database search results

### ISSUE THREAD

**Issue by thesunshade** [suttacentral/suttacentral#2997]
[…unless it's a literal search?]

#### Description:

Punctuation in search term keywords seems to have an effect on how the result is shown

#### Steps to reproduce:

* Search for [`sheepskins, goatskins`](https://suttacentral.net/search?query=sheepskins,%20goatskins) (Note the comma. It's there because I copy pasted from an actual text)

Every thing looks good:
![image](https://github.com/suttacentral/suttacentral/assets/82448383/c92553a0-537c-47a4-9141-52044a44a04f)
(ignore the fact that the [search terms are out of order](https://github.com/suttacentral/suttacentral/issues/2996))

This time…
* Search for [`sheepskins goatskins`](https://suttacentral.net/search?query=sheepskins%20goatskins) (note lack of comma)

Now the result looks different:
![image](https://github.com/suttacentral/suttacentral/assets/82448383/97376a82-1afd-4e43-8b0a-1ffbec419448)

All of the excerpts are missing. 

Perhaps the search itself is told to ignore punctuation but the code that processes the excerpts is not?

#### Steps to reproduce # 2:

* Search for [`I don't understand`](https://suttacentral.net/search?query=I%20don%27t%20understand)

Nothing will be found because it is expecting I don’t understand with the smart/curly quote.
Expected behaviour:

Whether you search for `I don't understand`, `I don’t understand`, or even `I dont understand` the results should be identical.

#### Expected behaviour:

It's good that punctuation is ignored for searches. However that same punctuation should not affect the way excerpts are produced.

**Comment by thesunshade** [comment]
As a side note, it appears that maybe not all punctuation is ignored? For example `another’s thoughts` gives results but `anothers thoughts` doesn't.

**Comment by khemarato** [comment]
Yes: punctuation should probably only be used by the query parser (e.g. "exact string" or -except -this)

**Comment by thesunshade** [comment]
> -except -this

Right. I don't think we can exclude terms with the minus sign (a la Google) at the moment but I think we should be able to. I'm not a fan of `NOT`.

**Comment by thesunshade** [comment]
Came up again [here](https://discourse.suttacentral.net/t/where-can-i-find-na-ca-so-na-ca-anno/34569?u=snowbird) on the forum.

**Comment by thesunshade** [comment]
@mtnadi, I wonder if this is a bug you would like to try and tackle. It's another one of those things where the current behaviour doesn't match expectations.

**Comment by mtnadi** [comment]
Ah, thanks for tagging me.  It's currently a black box in my understanding of the code, but I shall investigate and see how much I can figure out.

**Comment by mtnadi** [comment]
Some initial thoughts and findings before I've gotten the full picture...

It seems there are two different scenarios behind the scenes, the first being the highlighting doesn't match the database search which corresponds to the first example.  The second being that there are different sorts of quotes, corresponding to the second example.

The first scenario may be related to this other bit of fun:
![screen](https://github.com/suttacentral/suttacentral/assets/155436111/e2b817ab-1241-4475-ba32-b49397e7208b)
or it may just be a separate issue, or related to #2981.  Two things: `AND` or `OR` operator in search causing almost duplicate highlighting as well as ordering still being funky.

Unsolicited two cents: I feel like a literal search ought to keep punctuation, purely from a me-as-a-user perspective, but I have no idea what usual conventions are for this, or what is possible behind the scenes...I'll see what it looks like.

**Comment by thesunshade** [comment]
> Unsolicited two cents: I feel like a literal search ought to keep punctuation, purely from a me-as-a-user perspective, but I have no idea what usual conventions are for this, or what is possible behind the scenes...I'll see what it looks like.

In general I agree that a literal search should not ignore punctuation. However since the corpus being searched is relatively small I doubt that in practice there will be that many cases where someone would need to narrow their search by including punctuation.

Also, the texts will have typographic punctuation like `”, “, —` etc. and most users won't be able to type those. So I think that it is better to just ignore them.

**Comment by mtnadi** [comment]
More initial thoughts - I see the enterprise edition of Arango has search highlighting.  I take it SC search highlight code is meant to attempt the same functionality.  But what this means is I'd wrap my head around it better to split it into the highlight code matching Arango's search behaviour (to some degree) on the one hand and dealing with punctuation on the other.  Otherwise it seems like fudging too many moving parts at once.  Absent any other background info, I'll see which is easier to tackle first!

**Comment by thesunshade** [comment]
Thanks! Are you in the Gitter/dev chatroom? It might be a quicker way to get any one0off questions answered by @ihongda

**Comment by mtnadi** [comment]
I am now...is there a specific room?  I found a suttacentral lobby, but somehow I don't think that's it.  But also, I'm generally wary of taking up paid developer time, so I'd default to only asking when I'm completely stuck until I have a better sense of how things are done around here :)

While I'm at it, I'm planning to stop programming for the vassa, but if I get somewhere with this ticket in the next couple weeks, I intend to continue until I can wrap up a somewhat contained change, else write out my findings.  Do give a shout if anything different would be helpful.

**Comment by thesunshade** [comment]
> I am now...is there a specific room?

Is it this?

https://matrix.to/#/#suttacentral_Dev:gitter.im

**Comment by ihongda** [comment]
Hi @thesunshade @mtnadi 🙏🏼 

>I see the enterprise edition of Arango has search highlighting. 

The Enterprise Edition does have this feature, but the Community Edition does not, so it needs to be handled manually, which does increase the complexity.

**Comment by ihongda** [comment]
![image](https://github.com/suttacentral/suttacentral/assets/4037966/5fcec855-8011-492e-a4db-0a367b83b272)


When calculating the highlighted segments, I removed most of the punctuation, which seems to work, but needs further testing.

**Comment by mtnadi** [comment]
That also works for the `sheepskins, goatskins` example above, great!  I did have this as one of my test cases though, and the highlight isn't picking up the results from DA 2, which have the two words in separate "segments" as well as in the same "segment" (whatever is considered a segment in a legacy text).
![screen](https://github.com/user-attachments/assets/57eca168-88bf-44ca-9cf6-6eab8011131b)

I have another (incomplete) potential workaround; it preserves punctuation in the displayed segment, but it's resulting in duplicate segments when the two search terms appear in the same segment.

**Comment by mtnadi** [comment]
I see this is closed, so this is in case it'd ever be helpful in future.  It's not looking like I'll have a full solution within a doable amount of time before I'd like to stop for the vassa, so I've pushed a different, gimpy solution to my [fork](https://github.com/mtnadi/suttacentral/tree/nadi-issue-2997) and you can do whatever you like, including absolutely nothing!, with it.

What it does: if arango search returns a result, then (so far in my testing) that result will have highlighted content.  That content is displayed with punctuation intact, and the clips that are highlighted don't appear more than once due to this code change.

What it doesn't do: not every word of the search term gets highlighted.  I haven't gotten to highlighting multiple terms in one clip (I think that's a different issue/task).  It also doesn't handle issues that belong within the realm of how the search term is passed to arango (see *'d test cases below).  I also didn't want to mess with quotes yet, so it currently handles only `[',', '.', '-', ';', ':', '?', '!']`

What I tested with:
`sheepskins, goatskins`
`sheepkins,goatskins`
`sheepskins goatskins`
\* `wisdom, unattached` <- arango itself returns a different set of results than for `wisdom AND unattached`
`wisdom AND unattached` <- this gives example of duplicate clips e.g. in result for Snp 1.9
`wisdom,unattached`
`wisdom unattached`
\* `unattached wisdom` <- interesting arango returns nothing, but I think this is a different issue
`rose-apple in:vinaya`

Slightly tangential: it seems arango will look for the terms (e.g.  for "unattached, wisdom") to appear within the same segment (or segment-like-thingy in non-segmented texts).  When given `AND` (e.g. "unattached AND wisdom), then it searches anywhere within the document.  I'm guessing this is already known behaviour to some or all devs here!

Unrelated interestingness: search completely breaks if double quotes are entered.  It looks akin to a case of SQL injection =D

**Comment by ihongda** [comment]
Because it is linked to the related PR, it will be automatically closed when the PR is merged.

I have reopened it.

**Comment by thesunshade** [comment]
I'm guessing this has introduced a bug where the full stops are removed from the segment numbers in the highlight segments. For example:

![image](https://github.com/user-attachments/assets/034f4b61-075d-492a-915e-e8825315db9a)

All of those segment numbers should have a decimal in them. The decimal is also missing in the link, so the link is broken.
